### PR TITLE
add asset render errors to each platform

### PR DESCRIPTION
**Pull request by tmarmer**
- Adding custom errors for any error that happens at render time on each platform. The purpose of these errors is to highlight the specific asset that threw without losing the context of the original error.

### Change Type (required)
Indicate the type of change your pull request is:

- [ ] `patch`
- [X] `minor`
- [ ] `major`
- [ ] `N/A`

### Does your PR have any documentation updates?
- [ ] Updated docs
- [ ] No Update needed
- [ ] Unable to update docs

### CODE REVIEW

File: react/player/src/asset/index.tsx [hunk 1]
```diff
@@ -90,5 +101,23 @@ export const ReactAsset = (
     );
   }
 
-  return <Impl key={unwrapped.id} {...unwrapped} />;
+  return (
+    <ErrorBoundary
```
[the remark this answers is not in the document]
Comment (tmarmer) [reply react/player/src/asset/index.tsx]: The purpose of doing it at the asset level was to narrow down the specific asset that threw the error. If we only catch at the highest level we won't have access to the asset object where the error was thrown. I think this will also be needed for when we implement render errors for the `ErrorController`

File: android/player/src/androidTest/kotlin/com/intuit/playerui/android/renderer/BrokenAssetTest.kt [hunk 2]
```diff
@@ -42,9 +44,16 @@ internal class BrokenAssetTest {
...
+        var thrown: Throwable? = null
+        try {
             BrokenAsset(baseContext.copy(asset = runtime.asset(shouldFail = true))).render(appContext)
+        } catch (exception: Throwable) {
+            thrown = exception
         }
+
+        assertNotNull(thrown)
+        assertTrue(thrown is AssetRenderException)
+        assertTrue((thrown as AssetRenderException).cause is StaleViewException)
```
[the remark this answers is not in the document]
Comment (sugarmanz) [reply android/player/src/androidTest/kotlin/com/intuit/playerui/android/renderer/BrokenAssetTest.kt]: Hmm, good q — the `StaleViewException` is really just internal, to let re-rendering occur when things are "stale". Which actually is a facet of asset-level rendering recovery, so funny that this is what we're tackling. Could `StaleViewException` extend `AssetRenderException` so that it maintains it's identity but still complies with the expected raised exception type?

I think, as-is, this breaks the recovery logic built-in [here](https://github.com/player-ui/player/blob/ea25ef2a63232ce97dc29957aa20d7581d94379a/android/player/src/main/kotlin/com/intuit/playerui/android/asset/RenderableAsset.kt#L135-L138).

**Comment by tmarmer** [comment]
Added some additional bench tests for when the asset does throw an error and saw there is some additional time when bubbling up:

No error boundary:
broke-no-boundary

With error boundary:
broke-with-boundary

Using a class component with a custom implementation for dealing with the errors does bring this down. Tried also updating `react-error-boundary` to latest but no change in results. I think there is some excessive logging with the error boundary that is slowing things down here.

File: android/player/src/androidTest/kotlin/com/intuit/playerui/android/renderer/BrokenAssetTest.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (tmarmer) [reply android/player/src/androidTest/kotlin/com/intuit/playerui/android/renderer/BrokenAssetTest.kt]: As far as I can tell, the recovery logic should still be fine. In `RenderableAsset` the try catch for all errors wraps around the recovery logic ([see here](https://github.com/player-ui/player/blob/b5283d4753d7f98e669b195063b108e9f883fa3e/android/player/src/main/kotlin/com/intuit/playerui/android/asset/RenderableAsset.kt#L134-L151)) so that should be able to catch first if there is a stale view. On `SuspendableAsset` there was already something catching the `StaleViewException` and failing the player state ([here](https://github.com/player-ui/player/blob/b5283d4753d7f98e669b195063b108e9f883fa3e/android/player/src/main/kotlin/com/intuit/playerui/android/asset/SuspendableAsset.kt#L73-L81))

I couldn't find any other edge cases to this, but happy to test anything you can think of.

File: android/player/src/main/kotlin/com/intuit/playerui/android/asset/SuspendableAsset.kt [hunk 3]
```diff
@@ -78,6 +78,13 @@ public abstract class SuspendableAsset<Data>(
                     exception,
                 ),
             )
+        } catch (exception: Throwable) {
+            if (exception is AssetRenderException) {
+                exception.addAssetParent(assetContext)
+                throw exception
```
[the remark this answers is not in the document]
Comment (sugarmanz) [reply android/player/src/main/kotlin/com/intuit/playerui/android/asset/SuspendableAsset.kt]: The proper way to account for this is to make the entire `render` lifecycle `suspend` — we almost started down this path a few months ago, but hit some roadblocks. This should be solved for organically once we've done that.

File: android/player/src/main/kotlin/com/intuit/playerui/android/asset/AssetRenderException.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package com.intuit.playerui.android.asset
+
+import com.intuit.playerui.android.AssetContext
+import com.intuit.playerui.core.player.PlayerException
+
+class AssetRenderException : PlayerException {
+    private val assetParentPath = mutableListOf<AssetContext>()
+
+    val rootAsset: AssetContext
+    override var message: String = ""
```
Comment (sugarmanz) [inline android/player/src/main/kotlin/com/intuit/playerui/android/asset/AssetRenderException.kt]: I'm tempted to suggest that the message builder logic should just be contained in lazy initializer, but I know it's sensitive to `assetParentPath` changes, so it can't be `lazy`. So, the next suggestion would be a getter to build the message on the fly. There could be a caching mechanism for changes to the `assetParentPath` for performance, but that feels like a lot of overhead.

There is another comment I was going to make about preferring immutable lists for `assetParentPath`, which means any modification would require setting `assetParentPath`. So, we could lean into that route, to make `assetParentPath` immutable with a custom setter that will update the `message` on every change. Functionally, it'd be equivalent to what you have here, but would flow through idiomatic constructs rather than bespoke function calls.

Really not a huge deal, but could likely result in a cleaner construct.

File: android/player/src/main/kotlin/com/intuit/playerui/android/asset/AssetRenderException.kt [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    val initialMessage: String
+
+    constructor(rootAsset: AssetContext, message: String, exception: Throwable? = null) : super(message, exception) {
```
Comment (sugarmanz) [inline android/player/src/main/kotlin/com/intuit/playerui/android/asset/AssetRenderException.kt]: `internal` constructor — the only thing that should be able to construct this exception is us.

File: ios/swiftui/Tests/types/assets/ControlledAssetTests.swift [hunk 6]
```diff
@@ -111,11 +111,37 @@ class ControlledAssetTests: XCTestCase {
...
+        var thrownError: Error?
+        
+        do {
+            try registry.decode(value)
+        } catch {
+            thrownError = error
+        }
+        
+        guard let err = thrownError else { return XCTFail("expected asset to throw error while decoding") }
+        XCTAssertTrue(err.playerDescription.contains("Exception occurred in asset with id 'someId' of type 'someType'"))
```
Comment (nancywu1) [inline ios/swiftui/Tests/types/assets/ControlledAssetTests.swift]: can we add a test that tests a nested asset maybe like 
```
let val = context.evaluateScript("""
            ({
                id: 'outerActionId',
                type: 'action',
                label: {
                    asset: {
                        id: 'innerActionId',
                        type: 'action',
                        label: {
                            asset: {
                                id: 'textId',
                                type: 'text',
                                valueToTriggerDecodeFailure: 123
                            }
                        }
                    }
                }
            })
        """)
        
 ```

File: plugins/console-logger/ios/Sources/PrintLoggerPlugin.swift [hunk 7]
```diff
@@ -28,21 +28,29 @@ public class PrintLoggerPlugin: NativePlugin {
     public init(level: LogLevel = .error) {
         self.logLevel = level
     }
+    
+    private func printMessage(level: LogLevel, items: [Any]) {
+        print("[Player] [\(level.description)]", terminator: "")
+        items.compactMap({ $0 }).forEach({ print($0, terminator: "") })
+        print()
```
Comment (nancywu1) [inline plugins/console-logger/ios/Sources/PrintLoggerPlugin.swift]: can we avoid multiple prints to make it cleaner?
```
    let message = items.compactMap { $0 }.map { "\($0)" }.joined()
    print("[Player] [\(level.description)]\(message)")
 ```

**Review by KetanReddy — APPROVED** [review]
